### PR TITLE
LPS-183918 add copy document option

### PR DESCRIPTION
**Pull request by marcogalluzzi**
See issue [LPS-183918](https://issues.liferay.com/browse/LPS-183918)

BTW: I currently have one JSP for the copy of both file entries and file shortcuts. I would leave for the future this refactoring: using the same JSP also for the copy of folder.

### CODE REVIEW

**Comment by liferay-lima** [comment]
ci:test:sf

**Comment by liferay-lima** [comment]
ci:test:relevant

**Comment by Hanlf** [comment]
ci:test:relevant

**Comment by Hanlf** [comment]
Hi @marcogalluzzi ,
There are some failures are related to this PR, could you take a look?
1. [portal-license-smoke-mysql57-jdk8/0/0](https://test-1-30.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/521675//consoleText) and [modules-compile-jdk8/0/1](https://test-1-17.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/207691//consoleText)
```
     [exec] * Exception is:
     [exec] org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':apps:document-library:document-library-web:compileJSP'.
```
2. [modules-semantic-versioning-jdk8/0/0](https://testray.liferay.com/reports/production/logs/2023-05/test-1-37/test-portal-acceptance-pullrequest(master)/11515/modules-semantic-versioning-jdk8/0/0/jenkins-console.txt.gz?authuser=0)
```
     [exec] com.liferay.semantic.versioning.SemanticVersioningTest > testSemanticVersioning[/apps/document-library/document-library-api] FAILED
     [exec]     java.lang.AssertionError:   PACKAGE_NAME                                       DELTA      CUR_VER    BASE_VER   REC_VER    WARNINGS  
     [exec]     = ================================================== ========== ========== ========== ========== ==========
     [exec]     * com.liferay.document.library.display.context       MINOR      9.0.0      9.0.0      9.1.0      VERSION INCREASE REQUIRED
     [exec]     	<   class      com.liferay.document.library.display.context.DLUIItemKeys
     [exec]     		+   field      COPY
     [exec]     			+   access     final
     [exec]     			+   access     static
     [exec]     			+   return     java.lang.String
     [exec]     [Baseline Warning] Bundle Version Change Recommended: 20.1.0
```
3. [PortalLogAssertorTest-modules-integration-mysql57-jdk8/2/22](https://testray.liferay.com/reports/production/logs/2023-05/test-1-37/test-portal-acceptance-pullrequest(master)/11515/modules-integration-mysql57-jdk8/2/22/jenkins-console.txt.gz?authuser=0), not sure about this one if it is related, but it failed on the two relevant jobs.
```
     [exec] com.liferay.oauth.client.test.LocalOAuthClientTest > test FAILED
     [exec]     java.lang.AssertionError
```
4. [service-builder-jdk8/0/0](https://testray.liferay.com/reports/production/logs/2023-05/test-1-37/test-portal-acceptance-pullrequest(master)/11515/service-builder-jdk8/0/0/jenkins-console.txt.gz?authuser=0), also not sure about this one, but it failed on the two relevant jobs.
5. [DMPermissions#EntryCanNotBeEditedWithoutPermissions](https://testray.liferay.com/zh/home/-/testray/case_results/2306965803) I guess it should be a test fix, but I am unable to build a bundle locally with this PR, got the same error as issue1, I will investigate further after you fix the issue about document-library-web

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFileDisplayContext.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+import com.liferay.portal.kernel.util.ParamUtil;
+
+import javax.portlet.RenderResponse;
+
+import javax.servlet.http.HttpServletRequest;
+
+/**
+ * @author Marco Galluzzi
+ */
+public class DLCopyFileDisplayContext {
```
Comment (adolfopa) [inline modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFileDisplayContext.java:47]: You're not copying files, but file entries and/or file shortcuts. Using "file" in the name implies that you're only moving bytes around.

As this works for both file entries and file shortcuts, you need to use something that refers to both things (e.g. CopyEntry, like in `EditEntryMVCActionCommand`).

**Review by adolfopa — CHANGES_REQUESTED** [review]

**Comment by marcogalluzzi** [comment]
ci:test:sf

**Comment by marcogalluzzi** [comment]
ci:test:relevant

File: modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFileDisplayContext.java (hunk 1 above)
Comment (marcogalluzzi) [reply modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/display/context/DLCopyFileDisplayContext.java]: Change done. I'm using "CopyEntry". Thanks for the suggestion.

**Comment by marcogalluzzi** [comment]
ci:test:sf

File: modules/apps/document-library/document-library-test/src/testIntegration/java/com/liferay/document/library/app/service/test/DLAppServiceWhenCopyingAFileShortcutTest.java [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+			sourceGroupId, DLFolderConstants.DEFAULT_PARENT_FOLDER_ID,
+			fileEntry.getFileEntryId(),
+			ServiceContextTestUtil.getServiceContext(sourceGroupId));
+
+		DLAppServiceUtil.copyFileShortcut(
+			fileShortcut.getFileShortcutId(), targetParentFolder.getFolderId(),
+			targetParentFolder.getGroupId(),
+			ServiceContextTestUtil.getServiceContext(
+				targetParentFolder.getGroupId()));
+	}
```
Comment (adolfopa) [inline modules/apps/document-library/document-library-test/src/testIntegration/java/com/liferay/document/library/app/service/test/DLAppServiceWhenCopyingAFileShortcutTest.java:158]: Same as my previous comment in the other test file. Remove duplication by adding a `copy(long, long, long, long)` method (and override for convenience if necessary).

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java [hunk 3]
```diff
@@ -716,6 +716,36 @@ public FileEntry checkOutFileEntry(
...
+		Repository sourceRepository = repositoryProvider.getFileEntryRepository(
+			fileEntryId);
+
+		return copyFileEntry(
+			getRepository(destinationRepositoryId),
+			sourceRepository.getFileEntry(fileEntryId), destinationFolderId,
+			serviceContext);
+	}
+
+	public FileShortcut copyFileShortcut(
```
Comment (adolfopa) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java:734]: Shouldn't there be an `@Override` here?

**Review by adolfopa — CHANGES_REQUESTED** [review]
Sorry there were a couple of things I didn't see in my previous review :pleading_face:

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java (hunk 3 above)
Comment (marcogalluzzi) [reply portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java:734]: Yes! Strange the source formatter doesn't complaint.

Comment (marcogalluzzi) [reply portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLAppServiceImpl.java:735]: done

**Comment by marcogalluzzi** [comment]
ci:test:sf

**Comment by marcogalluzzi** [comment]
ci:test:relevant

**Review by adolfopa — APPROVED** [review]

**Comment by marcogalluzzi** [comment]
Hi @Hanlf please have a look to the failing tests of this PR and forward to Brian if you think they aren't related (I think they are not). Thanks.

**Comment by marcogalluzzi** [comment]
ci:reevaluate:1277401_8216

**Comment by marcogalluzzi** [comment]
ci:test:stable

**Comment by marcogalluzzi** [comment]
ci:test:stable

**Comment by Hanlf** [comment]
This test [PortalLogAssertorTest-modules-integration-mysql57-jdk8/2/22](https://testray.liferay.com/reports/production/logs/2023-05/test-1-26/test-portal-acceptance-pullrequest(master)/9254/modules-integration-mysql57-jdk8/2/22/jenkins-console.txt.gz?authuser=0) fails continuously in this PR.
Re-run it to make sure.

**Comment by Hanlf** [comment]
ci:test:relevant

**Comment by Hanlf** [comment]
Hi @marcogalluzzi 
I need your help to check if the failure [PortalLogAssertorTest-modules-integration-mysql57-jdk8/2/22](https://testray.liferay.com/reports/production/logs/2023-05/test-1-26/test-portal-acceptance-pullrequest(master)/9269/modules-integration-mysql57-jdk8/2/22/jenkins-console.txt.gz?authuser=0) is relevant, from the [test history](https://testray.liferay.com/zh/home/-/testray/cases/2219048214), the error `PortalLogAssertorTest#testScanXMLLog: null java.lang.NullPointerException  at com.liferay.oauth2.provider.rest.internal.endpoint.liferay.LiferayOAuthDataProvider._populateClient` only occurs on this PR :(

**Comment by marcogalluzzi** [comment]
Hi @Hanlf I've just run the failing (and IMO unrelated) test `LocalOAuthClientTest` with last master (without my changes) and it also fails! Having a look to the remaining fails, do you think is OK to forward? Thanks

**Comment by Hanlf** [comment]
Thanks for your checking, then I think it is ready to go!

**Comment by Hanlf** [comment]
ci:forward:force

**Pull request closed without merging**